Send notifications when there's a new dialogue form for you to fill in

lagerros opened the pull request:
I think there's something like ~30-40 users who haven't replied to their partners matching form.

This adds a cronjob that checks for this situation, and if it finds it, pings the user to fill in the form. 

In future we can easily update it to work with a callback instead, but doing the cron for now to inform all of these users.

<img width="167" alt="image" src="https://github.com/ForumMagnum/ForumMagnum/assets/21347592/f7e4f40a-9173-42b6-a2b5-d5acb5d483f4">

b0b3rt commented on packages/lesswrong/components/users/DialogueMatchingPage.tsx:
```diff
@@ -931,6 +933,18 @@ const NextStepsDialog = ({ onClose, userId, targetUserId, targetUserDisplayName,
     setTopicNotes('')
   }
 
+  const onCloseWrapper = () => {
+    // used to remove query parameters if user clicks outside the dialog
+    const searchParams = new URLSearchParams(location.search);
+    searchParams.delete('dialogueCheckId');
+
+    const newLocation = location.pathname + '?' + searchParams.toString();
+
+    navigate(newLocation);
+  
+    onClose();
+  };
```
What is this doing, functionally?

(Regardless of what it's doing, that's not how to clear query params from the URL with a navigation - pretty sure all you need to do is `navigate({ ...location, search: '' })`, where `location` is the entire object coming back from `useLocation`, rather than the `location` field inside of that object.)

b0b3rt replied:
Nvm, figured it out from the notification.

lagerros replied:
The purpose of this code: if you don't have this then, when you're on the dialogueMatching page and watching the form, if you click in the whitespace outside the form, it will not close. That's undesired! 

In addition, I didn't want to clear out all query parameters, because maybe later someone else will stuff some query parameters on that page, and I really only want to clear the one I'm using -- dialogueCheckId -- and not interfere with anyone else

b0b3rt commented on packages/lesswrong/components/users/DialogueMatchingPage.tsx:
```diff
@@ -1430,6 +1444,8 @@ export const DialogueMatchingPage = ({classes}: {
 
   const updateCurrentUser = useUpdateCurrentUser()
   const currentUser = useCurrentUser();
+  const { location, query, params } = useLocation()
```
Is this being used for anything?

b0b3rt replied:
Oh, I see it's being used for the `dialogueCheckId` query.  Are the other two things used anywhere?

lagerros replied:
nope, removing

b0b3rt commented on packages/lesswrong/server/dialogues/cron.ts:
```diff
@@ -25,6 +26,30 @@ addCronJob({
+      void createNotification({
+        userId: check.userId,
+        notificationType: "yourTurnMatchForm",
+        documentType: null,
+        documentId: null,
```
We have a `dialogueCheck` document type, which I'm pretty sure can be appropriately used here

b0b3rt replied:
And then you don't need to pass `checkId` and `targetUserId` through `extraData`.

b0b3rt commented on packages/lesswrong/lib/notificationTypes.tsx:
```diff
@@ -402,6 +402,27 @@ export const NewDialogueCheckNotification = registerNotificationType({
+    return <DebateIcon style={iconStyles}/>
+  },
+  getLink({ extraData }: { extraData: Record<string,any> }) { 
+    return `/dialogueMatching${extraData?.checkId ? '?dialogueCheckId='+extraData?.checkId : ''}`
```
Please build this path using `URL`

b0b3rt commented on packages/lesswrong/components/users/DialogueMatchingPage.tsx:
```diff
@@ -1506,6 +1537,20 @@ export const DialogueMatchingPage = ({classes}: {
 
   const prompt = "Opt-in to LessWrong team viewing your checks, to help proactively suggest and facilitate dialogues" 
 
+  if (!!activeDialogueCheck && !!activeTargetUser) {
```
I was confused by what this was supposed to be doing based on the naming of `activeDialogueCheck` and `activeTargetUser` until I figured out what the whole PR was doing.  Would rename

b0b3rt commented on packages/lesswrong/components/users/DialogueMatchingPage.tsx:
```diff
@@ -1506,6 +1537,20 @@ export const DialogueMatchingPage = ({classes}: {
+    openDialog({
+      componentName: 'NextStepsDialog',
+      componentProps: {
+        userId: currentUser?._id,
```
Isn't `currentUser` guaranteed to not be null by this point?

b0b3rt commented on packages/lesswrong/server/repos/UsersRepo.ts:
```diff
@@ -572,6 +572,7 @@ class UsersRepo extends AbstractRepo<"Users"> {
   
     return result;
   }
+
```
remove extraneous diffs

b0b3rt commented on packages/lesswrong/server/repos/DialogueChecksRepo.ts:
```diff
@@ -58,6 +63,32 @@ class DialogueChecksRepo extends AbstractRepo<"DialogueChecks"> {
+        dc.*,
+        dmp_reciprocal._id AS "targetUserMatchPreferenceId"
+      FROM "DialogueChecks" AS dc
+      LEFT JOIN "DialogueMatchPreferences" AS dmp ON dc._id = dmp."dialogueCheckId"
+      LEFT JOIN
+          "DialogueChecks" AS dc_reciprocal
+          ON
+              dc."userId" = dc_reciprocal."targetUserId"
+              AND dc."targetUserId" = dc_reciprocal."userId"
+      LEFT JOIN "DialogueMatchPreferences" AS dmp_reciprocal ON dc_reciprocal._id = dmp_reciprocal."dialogueCheckId"
+      LEFT JOIN "Notifications" AS n ON n."extraData"->>'targetUserMatchPreferenceId' = dmp_reciprocal._id
```
The formatting here is making it pretty difficult to read.  Generally prefer:
- `ON` on new lines after the `JOIN`
- the table & alias on the same line as the `JOIN`

b0b3rt commented on packages/lesswrong/server/dialogues/cron.ts:
```diff
@@ -25,6 +26,30 @@ addCronJob({
 });
 
+
+addCronJob({
```
Actually, if you're going to handle this via callbacks going forward, but just want a one-time backfill, you don't want to use a cron job for that, just a `manualMigration`.

lagerros replied:
I say we could, not that we necessarily would. For now it's built this way, so I'd say let's just ship and maybe one glorious day it'll be callbacked?

b0b3rt commented on packages/lesswrong/server/repos/DialogueChecksRepo.ts:
```diff
@@ -58,6 +63,32 @@ class DialogueChecksRepo extends AbstractRepo<"DialogueChecks"> {
+              dc."userId" = dc_reciprocal."targetUserId"
+              AND dc."targetUserId" = dc_reciprocal."userId"
+      LEFT JOIN "DialogueMatchPreferences" AS dmp_reciprocal ON dc_reciprocal._id = dmp_reciprocal."dialogueCheckId"
+      LEFT JOIN "Notifications" AS n ON n."extraData"->>'targetUserMatchPreferenceId' = dmp_reciprocal._id
```
I am also pretty unsure that the way this query is constructed is right, or will perform reasonably, given that `Notifications` is a big table and `extraData` isn't indexed at all.  If you do [this](https://github.com/ForumMagnum/ForumMagnum/pull/8548/files#r1433352642), you'll be able to join on `documentId`.

Separately, I don't see `targetUserMatchPreferenceId` being used anywhere in the code after being passed in to the notification?  If it's not needed, then my guess is that most of this query goes away (and maybe can actually just be an existing query that we already have for finding matched dialogue checks).

lagerros replied:
In testing it runs instantly. (I build queries using Hex so tend to get exposed to them being slow)

`targetUserMatchPreferenceId` is used... wait for it... in the JOIN in this query itself! 
Because what we're doing is notifying users about a new form to pay attention to. And the id of that form is a pointer to the new ground truth object we're trying to notify them about.

lagerros commented:
Shipped a rewrite. Updates:
* rewritten to use callback
* removed cron
* add manual migration to backfill previous notifications

Tested in dev for: 
* Callback sending new notifications when it should (and to only the users it should)
* Adjacent form functionality (creating new dialogues, etc) still working
* Manual migration creates new notifications
* Manual migration has idempotent behavior

b0b3rt commented on packages/lesswrong/lib/notificationTypes.tsx:
```diff
@@ -164,6 +166,16 @@ export const getDocumentSummary = async (documentType: NotificationDocument | nu
         displayName: null,
       }
+    case 'dialogueMatchPreference':
+      const dialogueMatchPreference = await DialogueMatchPreferences.findOne({ _id: documentId })
+      const dialogueCheckMatch = await DialogueChecks.findOne({ _id: dialogueMatchPreference?.dialogueCheckId })
+      const userMatch = await Users.findOne(dialogueCheckMatch?.userId)
```
I would rewrite this so that we potentially passing in `undefined` for the _id selectors to `DialogueChecks.findOne` and `Users.findOne` (if either `dialogueMatchPreference` or `dialogueCheckMatch` return `null`).  That'll pull up a random document of that type.

Something like:
```
      const dialogueMatchPreference = await DialogueMatchPreferences.findOne(documentId);
      if (!dialogueMatchPreference) return null;

      const dialogueCheckMatch = await DialogueChecks.findOne(dialogueMatchPreference.dialogueCheckId);
      if (!dialogueCheckMatch) return null;

      const userMatch = await Users.findOne(dialogueCheckMatch.userId);
      if (!userMatch) return null;

      return { ... };
```

b0b3rt replied:
This is more robust against any potential future changes which might allow random user's dialogue check info to slip through in notifications for this document type.

b0b3rt commented on packages/lesswrong/server/callbacks/dialogueMatchPreferenceCallbacks.ts:
```diff
@@ -190,8 +190,7 @@ const welcomeMessage = (formDataSourceUser: MatchPreferenceFormData, formDataTar
+const getReciprocalMatchPreference = async (context: ResolverContext, currentUser: DbUser | null, dialogueCheckId: string) => {
   const dialogueCheck = await context.loaders.DialogueChecks.load(dialogueCheckId);
 
   // This shouldn't ever happen
```
FYI the line under this looks like it's using `??` instead of the `||` operator, where it needs to be using `||` (since it's checking whether any of the conditions evaluate to `true`, and `??` won't do anything when operating on booleans, which those will be since they're all either coerced with `!` or coming out of equality comparisons).

lagerros replied:
This is code written by Oli.

b0b3rt commented on packages/lesswrong/server/callbacks/dialogueMatchPreferenceCallbacks.ts:
```diff
@@ -275,3 +282,21 @@ getCollectionHooks("DialogueMatchPreferences").createBefore.add(async function G
   return userMatchPreferences;
 });
+
+getCollectionHooks("DialogueMatchPreferences").createAfter.add(async function NotifyUsers ( userMatchPreference, { context, currentUser } ) {
```
I'd maybe give the function a more descriptive name, in case it ever comes up in any stack traces

b0b3rt commented on packages/lesswrong/server/repos/DialogueChecksRepo.ts:
```diff
@@ -58,6 +61,28 @@ class DialogueChecksRepo extends AbstractRepo<"DialogueChecks"> {
+          AND (dmp._id IS NULL OR dmp.deleted IS TRUE)
+          AND dmp_reciprocal._id IS NOT NULL
+          AND dmp_reciprocal.deleted IS NOT TRUE 
+          AND (n._id IS NULL OR n.type != 'yourTurnMatchForm')
```
Actually, now that I think about it, this last condition seems like it might cause us to incorrectly return dialogue checks with a `targetUserMatchPreferenceId` in the case where we've actually sent a notification for that dialogue match preference of any type other than `yourTurnMatchForm`, even if we _have_ also sent a `yourTurnMatchForm` notification for that dialogue match preference?  And the reason it's not a problem right now is that this is the only notification we have for that docment type (so far).

I don't think it's worth the effort to fix it, but to avoid this being a future footgun I'd feel a lot better if this query was just run inline in the manual migration (see `2023-07-28-viewUpdaterErrorHandling.ts` for an example of how to do this) rather than being defined as a method on the repo, with a comment above this line explaining this in case anyone is tempted to copy the query for other purposes.

lagerros replied:
That is intended behavior. 

The intended behavior is not to send only one notification of any kind per matchPreference. It's to only send one `yourTurnMatchForm` per matchPreference -- because there might be other things we want to inform users of!

b0b3rt reviewed (approved):
lgtm
